File: interpreter.go
/**
  @MDDemirci
**/


package main

import "os"
import "fmt"
import "io/ioutil"
import "reflect"

func in_array(val interface{}, array interface{}) (exists bool, index int) {
    exists = false
    index = -1

    switch reflect.TypeOf(array).Kind() {
    case reflect.Slice:
        s := reflect.ValueOf(array)

        for i := 0; i < s.Len(); i++ {
            if reflect.DeepEqual(val, s.Index(i).Interface()) == true {
                index = i
                exists = true
                return
            }
        }
    }

    return
}

func check(e error) {
    if e != nil {
        fmt.Println("Error : %s",e)
    }
}

func read_SourceCode(file_name string){
    dat, err := ioutil.ReadFile(file_name)
    check(err)
    interpreter(string(dat))
}

func stringToArray(sourceCode string) []string{
  array := []string{}
  for _, r := range sourceCode {
      c := string(r)
      array = append(array,c)
  }
  return array
}

func interpreter(sourceCode string) {
  i := 0
  RAM := []int{0}
  indexInRAM := 0
  whileArray := []int{}

  codeArray := stringToArray(sourceCode)
  for i < len(codeArray) {
    if codeArray[i] == "["{
      inArray,_ := in_array(i+1,whileArray)
      if !inArray{
        whileArray = append(whileArray,i+1)
      }
    }else if codeArray[i] == "]"{
      if RAM[indexInRAM] == 0{
          whileArray = whileArray[:len(whileArray)-1]
      }else {
        i = whileArray[len(whileArray)-1]
        continue
      }
    }else if codeArray[i] == ">"{
      indexInRAM = indexInRAM + 1
      if len(RAM) >= indexInRAM{
        RAM = append(RAM,0)
      }
    }else if codeArray[i] == "<"{
      indexInRAM = indexInRAM - 1
    }else if codeArray[i] == "+"{
      RAM[indexInRAM] = RAM[indexInRAM] + 1
    }else if codeArray[i] == "-"{
      RAM[indexInRAM] = RAM[indexInRAM] - 1
    }else if codeArray[i] == "."{
      fmt.Printf("%c\n", RAM[indexInRAM])
    }else if codeArray[i] == ","{
      // TO DO
    }
    i += 1
  }
}

func main() {
    // `os.Args` provides access to raw command-line
    // arguments. Note that the first value in this slice
    // is the path to the program, and `os.Args[1:]`
    // holds the arguments to the program.

    argsWithoutProg := os.Args[1:]

    if len(argsWithoutProg) == 1 {
      read_SourceCode(argsWithoutProg[0])
    }else {
      fmt.Println("go run interpreter.go <inputfile>")
    }
}

File: interpreter.swift
// @MDDemirci

import Cocoa

extension String {

    var length: Int {
        return self.characters.count
    }

    subscript (i: Int) -> String {
        return self[Range(i ..< i + 1)]
    }

    func substring(from: Int) -> String {
        return self[Range(min(from, length) ..< length)]
    }

    func substring(to: Int) -> String {
        return self[Range(0 ..< max(0, to))]
    }

    subscript (r: Range<Int>) -> String {
        let range = Range(uncheckedBounds: (lower: max(0, min(length, r.lowerBound)),
                                            upper: min(length, max(0, r.upperBound))))
        let start = index(startIndex, offsetBy: range.lowerBound)
        let end = index(start, offsetBy: range.upperBound - range.lowerBound)
        return self[Range(start ..< end)]
    }

}

extension UInt8 {
    var char: Character {
        return Character(UnicodeScalar(self))
    }
}

extension Int {
    var char: Character {
        return Character(UnicodeScalar(self)!)
    }
}

func isInArray(value: Int, array: [Int]) -> Bool {

    for a in array {
        if a == value {
            return true
        }
    }
    return false
}


func interpreter(sourceCode: String) {
    var i = -1
    var RAM: [Int: Int] = [0: 0]
    var indexInRAM = 0
    var whileArray: [Int] = []
    while i < sourceCode.length && i <= 100 {
        i += 1
        switch sourceCode[i] {
        case "[":
            if !isInArray(value: i, array: whileArray) {
                whileArray.append(i)
            }
        case "]":
            if RAM[indexInRAM] == 0 {
                whileArray.removeLast()
            } else {
                i = whileArray.last!
            }
            continue
        case ">":
            indexInRAM = indexInRAM + 1

            if RAM[indexInRAM] == nil { //nil check
                RAM[indexInRAM] = 0
            }
        case "<":
            indexInRAM = indexInRAM - 1
        case "+":
            if let r = RAM[indexInRAM] { //nil check
                RAM[indexInRAM] = r + 1
            }
        case "-":
            RAM[indexInRAM] = RAM[indexInRAM]! - 1
        case ".":
            if let r = RAM[indexInRAM] {
                print(r.char)
            }
        case ",":
            //TODO input
            break
        default:
            break
        }

    }
}

func readFile(name: String) -> String? {

    if let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {

        let path = dir.appendingPathComponent(name)

        //reading
        do {
            let text = try String(contentsOf: path, encoding: String.Encoding.utf8)
            return text
        }
        catch { /* error handling here */ }
    }
    return nil
}

let name = ProcessInfo.processInfo.arguments[1]
let text = readFile(name: name)
if let t = text {
    interpreter(sourceCode: t)
}



